### ISSUE TITLE
[GJEP] `@dist` DSL for GenJAX

### ISSUE THREAD

**Issue by femtomc** [genjax-community/genjax#227]
## Background

As I've been working with @Joaoloula, Joao has suggested several proposals to extend the combination patterns for the `Distribution` class in GenJAX.
For reference, `Distribution` supports the following interfaces:

```python
@dataclass
class Distribution(GenerativeFunction):
    @abc.abstractmethod
    def random_weighted(self, *args, **kwargs):
        pass

    @classmethod
    @abc.abstractmethod
    def estimate_logpdf(cls, key, v, *args, **kwargs):
        pass
```

> These are not the normal interfaces that you'd expect for _exact densities_ - but they are the lower level interfaces which `Prox` fits within. `ExactDensity` is a subtype of `Distribution` which implements exact density semantics for these methods.

Our conversations lead me to consider the existing means of combination of distributions (the only one is `Product` -- but there's no mixture combination patterns, for example).

One very general means of combining distributions is an `@dist` DSL which supports the ability to transform samples from distributions using functions which support inverse log determinant Jacobian computations. Currently GenJAX doesn't provide such a DSL. 

The purpose of this GJEP issue is to explore the design of such a DSL within GenJAX.

## Proposal

A generative function language which supports a function-like specification pattern, whose key operation elements include:
1. Sampling from primitive distributions.
2. Transforming the samples with a function $f$ that supports ILDJ computations.

would be useful - and the language itself can likely support more expressive patterns than [Gen.jl's `@dist`][gen_dist] DSL by virtue of our control for JAX program transformations.

To implement the GFI for programs expressed in this DSL, we'd construct a class:

```python
class DistributionTransformDSLGenerativeFunction
    source: Callable
    ...
```

Creating instances of this generative function type would require that the user use a function-like DSL (I've hesitantly changed `@dist` to `@trans` pending a review of the expressive power of this language compared to the `@dist` DSL).

```python
@trans
def transformed_dist(*args):
    x = genjax.random_variable(genjax.Normal)(0.0, 1.0)
    return some_ildj_compat_fn(x)
```

Here, `genjax.random_variable` is a primitive which works inside this DSL and with the transformations which we hope to define to support ILDJ computations.

The ILDJ correction term semantics would be described by a function transformation - which would require some form of inversion.

At this stage, it's good to consider further design background for a system called [Oryx][oryx] which supports a DSL and transformations which are close to (if not exactly) what we'd be looking to implement.

### Oryx background

Outside of this issue, I've discussed the design of another system called [Oryx][oryx] with @alex-lew - Oryx provides such a DSL, but it's more general than Gen.jl's DSL.

The design philosophy of Oryx is slightly different than Gen. Oryx is focused on providing function transforms and primitives which users can use inside of pure Python functions - it doesn't support "class-like" abstraction (like generative functions) but assumes users will just be working with transformations on Python functions.

Oryx and GenJAX rely on common transformation abstractions - like staging stateful effect handling to support inserting/recording state from pure functions. Oryx's `sow` and `reap` transformations are used by Oryx to support `log_prob` evaluation, and could also be used to support default proposal importance sampling with a bit of work. 

Roughly, we use the same type of transformations to support stateful GFI methods - but our interfaces work at a higher level of inference abstraction (e.g. we provide `importance`, which provides a pattern of `sow`/`reap` which we expect to be common when using custom proposals).

It's also not clear to me how compositionality works at the inference interface level with Oryx -- e.g. when we implement `simulate` for our "function transformation" DSL (the `BuiltinGenerativeFunction` DSL), the function transform uses `simulate` for generative function callees. I don't think Oryx has a notion of _inference compositionality_ yet - but I suspect if they pursued it, it would end up being close to what Gen prescribes.

That was a bit of background - but I haven't discussed the main feature from Oryx that I'm considering for the DSL - the `ildj` transformation and the `inverse` transformation.

#### Interpreters for ILDJ

In `Oryx` core - there's an interesting type of interpreter called a propagation interpreter. [This interpreter is also present in GenJAX](https://github.com/probcomp/genjax/blob/main/src/genjax/_src/core/interpreters/graph.py) - and in previous revisions of GenJAX, it was used to support incremental computation in the `update` GFI method (now `update` uses `CPS`, for somewhat obvious compile-time control flow reasons). 

> As far as I can tell - this interpreter was original due to Roy Frostig, and a modified/cleaned up version ended up in Oryx.

This interpreter works by wrapping JAX tracer values in a class type called `Cell` - which represents "JAX tracer values with a bit of extra information". `Cell` is philosophically similar to lattice value types in abstract interpretation. 

The propagation interpreter supports a partial concrete/abstract interpretation process - it treats the `Jaxpr` of a JAX compat function as a graph - where each register value (e.g. a `jax.core.Variable`) is a node. Each primitive instruction is an edge.

The propagation interpreter first sets the "in nodes" (e.g. the argument nodes into the `Jaxpr` call) and the "out nodes" to the `bottom` lattice value, and also creates nodes for constants (with `top` lattice value). Then, the propagation interpreter repeatedly performs a graph traversal - attempting to update the lattice information using primitive rules. When an update occurs, edges are added to a queue -- this process iterates until an information fixed point is achieved.

In Oryx, `inverse` and `ildj` are implemented using this interpreter. Oryx also provides customization interfaces for the propagation rules of the interpreter, supporting custom inverses (for example).

### Should we just use Oryx for the DSL?

By virtue of supporting the transformations and interfaces required for a distribution DSL that supports ILDJ compat function transformations of RVs - Oryx is an excellent candidate to provide support for this functionality.

However, I'm slightly concerned about using Oryx (for a few reasons).

**Ambiguous maintainership/support** I think Sharad Vikram was the lead dev and core maintainer of Oryx - but I'm not sure if that's true anymore. I have no insight into the internal Google processes surrounding Oryx. Oryx doesn't have an active open source community - so the significant contributors/merge authors are all Google employees. Oryx hasn't received significant updates in quite awhile (outside of being moved to the `jax-ml` org, and some recent `pjit` updates). It's not clear to me if Oryx will get a new maintainer.

**Oryx extras which we don't necessarily need** Oryx provides inference modules which build on top of its core transformation interfaces. One such module is `mcmc`. If we used Oryx, we'd just use our own versions of `mcmc` through the GFI - which support generic generative functions, and our versions also support incremental computation (this is missing from Oryx right now). I suspect that our versions are strictly more general - because of the GFI abstraction - Oryx's `mcmc` module can only be used with Oryx primitives/transformations.

There's positive reasons to consider Oryx as a source of new ideas. One important idea might be a learnable `@dist` DSL which includes neural network components. As far as I can tell, that's a major value proposition of Oryx - and there's parts of the `ILDJ` transformation (which I currently don't fully understand) that seemed like they are aimed at supporting neural network based function transformations of RVs.

This is a neat idea - and it would help us support deep generative models as distribution components in Gen programs.

## Unanswered proposal questions

How does the DSL I've described interact with `Prox`? This is probably easy to see for someone (maybe @alex-lew) - it's not clear to me what happens if log probs become estimates. I'd have to sit down and think through a few simple transformation examples. It's also not clear to me how taking a `ChoiceMap` return value and then transforming it with a function `f` would work -- or what examples of this are useful when there's internal structure to the `ChoiceMap` (in contrast, to `Marginal`, say -- where the return value is just the value of a single address).

The cool thing would be if it "just works" - you correct estimates with the ILDJ correction and the result is also an unbiased estimate. I don't know the answer to this right now -- but it's worth investigating and gaining a clear answer.

I should say I think the DSL described in this issue has high value irrespective of how it interacts with other languages (like Prox).

## Implementation recommendation

For implementation in the short term, I recommend pulling Oryx's core ILDJ transformation code into a new module in GenJAX `genjax.generative_function.distributions.trans`. The propagation interpreter is already part of GenJAX `core` - we can re-use Oryx's rule system and core functionality (with proper Apache attribution).

The implementation of `DistributionTransformDSLGenerativeFunction` will lightly wrap the transformations from Oryx.

In the future, if Oryx gets a new maintainer - we can just use Oryx as a dependency, and discuss package distribution patterns with the new maintainers (e.g. if we don't want `mcmc`, or `optimizers`, etc).

[gen_dist]: https://www.gen.dev/docs/stable/ref/distributions/#dist_dsl-1
[oryx]: https://github.com/jax-ml/oryx
[sharad]: https://sharadvikram.com/

**Comment by femtomc** [comment]
I'm discussing the maintenance model of Oryx (and questions about usage) with Sharad here: https://github.com/jax-ml/oryx/issues/38

**Comment by femtomc** [comment]
I spent some time last night and this morning mangling Oryx to understand the core transforms and implementation patterns. I was able to extract the core functionality and support a DSL which looks like the following:

<img width="601" alt="image" src="https://user-images.githubusercontent.com/34410036/218820477-e6b3aa36-d942-4eee-b95f-4622d7172aed.png">

There are some interesting issues with `log_prob` implementation of Oryx (ILDJ via transformation) (@alex-lew pointed this out originally) -- which raises some silent failures when using a language with Oryx's implementation.

<img width="601" alt="image" src="https://user-images.githubusercontent.com/34410036/218820662-f2cadbe1-5b1a-483a-8713-e15b11cd3ecb.png">

The Jacobian is not square — but this fails silently because the propagation interpretation pattern will essentially compute the ILDJ correction at the random variable for only one of the return values at random! (E.g. only one of the return values is required to have complete information about the value constraint at the random_variable call in the propagation interpreter pattern).

The implementation of Oryx is pretty complex — it took me about 7 hours to pull out the core inverse + ildj pieces … and Oryx doesn’t even use the `ildj` transform to implement `log_prob` — it uses the ildj primitive rules, but computes `log_prob` using a lower level interpreter implementation which I think are designed to improve efficiency. 

I wonder if there’s a conceptually more elegant implementation which avoids silently failing on these examples.

I was considering the `@dist` language framing as defining objects which have two functions:

```
m: A -> D Tau
f: Tau -> B
```

If I have `inverse_and_ildj` as a transform, I can take `inverse_and_ildj(f)` to get a correction + the inverse, and then use `m` to score the inverse and add the correction.

I think functionally, this is what we want to do. Oryx’s implementation departs from this pattern — but I think this pattern also allows us to check that `f` actually supports `ildj `— and perhaps we can reject programs like the above.

**Comment by alex-lew** [comment]
@femtomc Nice work! Excited to understand better what you've learned about Oryx, and what you've implemented. 

Some comments:

> There's positive reasons to consider Oryx as a source of new ideas. One important idea might be a learnable @dist DSL which includes neural network components. As far as I can tell, that's a major value proposition of Oryx - and there's parts of the ILDJ transformation (which I currently don't fully understand) that seemed like they are aimed at supporting neural network based function transformations of RVs.

Note that `@dist` DSL should also support gradients, especially with Xuan's recent bug fix!

> the language itself can likely support more expressive patterns than [Gen.jl's @dist](https://www.gen.dev/docs/stable/ref/distributions/#dist_dsl-1) DSL by virtue of our control for JAX program transformations.

Interesting, I'd be curious to see examples of this. As it stands, `@dist`'s design should be able to handle bijective transformations, and also some that are not bijective -- e.g., `@dist uniform_from_list(l) = l[uniform_discrete(1, length(l))]`. (This is not bijective because the list might contain repeated elements -- so there isn't a 1-1 mapping between indices and values.)

For low-dimensional (but still multivariate) invertible transformations, like `(x, y) -> (x, x - y)`, a "triangular" inversion strategy like that used by Stochaskell would be more powerful than the `@dist` DSL is. I think this may be how Jax's program inversion works too? But I am somewhat skeptical of this because it is a bit brittle -- every version I've seen of the idea breaks for `(x, y) -> (x + x, y + y)`, and I believe some break for `(x, y) -> (x + y, x - y)`. So small syntactic variations can break code for somewhat opaque reasons.

> I was able to extract the core functionality and support a DSL which looks like the following:

Nice!  But are you sure that you're getting correct densities for this example? `jnp.cos` is *not invertible*: the same output can arise from infinitely many different inputs. As such, the density of the transformed distribution is actually an *infinite sum* of ILDJ-corrected normal densities, evaluated at all the different values where `cos(x) = v`. This should be intractable, so I'm skeptical that the inversion is being handled properly. Note that with the Prox interfaces, it is valid to just *importance sample* a single value from the preimage, and compute an importance weight.

Re: syntax, my personal aesthetic preference would be to try to avoid things like `genjax.rv(genjax.Normal)(x, 1.0)` -- I think it's slightly non-Pythonic. But I don't understand all the design constraints arising from your implementation strategy.  I am also not sure about `@trans` or `@trans_dist` -- I don't think it's obvious what `trans` expands to (translate? transform?). I do like that you an directly call deterministic functions on the value `v`.


> How does the DSL I've described interact with Prox? This is probably easy to see for someone (maybe @alex-lew) - it's not clear to me what happens if log probs become estimates. 

In general it should be fine. This is why Prox still works with involutive MCMC in Gen, e.g.. If I want to estimate `correction * x` and I have an unbiased estimate `y` of `x`, then `correction * y` is an unbiased estimate of `correction * x`.

**Comment by femtomc** [comment]
@alex-lew 
> Re: syntax, my personal aesthetic preference would be to try to avoid things like genjax.rv(genjax.Normal)(x, 1.0) -- I think it's slightly non-Pythonic. But I don't understand all the design constraints arising from your implementation strategy. I am also not sure about @trans or @trans_dist -- I don't think it's obvious what trans expands to (translate? transform?). I do like that you an directly call deterministic functions on the value v.

I think I should be able to support syntax like the following:

```python
@genjax.dist
def new_dist(x):
    v =~ genjax.Normal(x, 1.0)
    return jnp.exp(v / 2.0) + 2.0
```

I'm trying not to introduce too many new syntactical notions. Another option would be:


```python
@genjax.dist
def new_dist(x):
    v = genjax.Normal(x, 1.0) @ RV
    return jnp.exp(v / 2.0) + 2.0
```

where `RV` is a special value (similar to `inline`, from PyGen).

I think it might be okay to leave overrides for `~` (denoting: a generative function call without an address).

It is quite annoying without macros to design something nice.

**Comment by femtomc** [comment]
Separately:

> Interesting, I'd be curious to see examples of this. As it stands, `@dist`'s design should be able to handle bijective transformations, and also some that are not bijective -- e.g., `@dist` uniform_from_list(l) = l[uniform_discrete(1, length(l))]. (This is not bijective because the list might contain repeated elements -- so there isn't a 1-1 mapping between indices and values.)

Here, I was just thinking that you can sample from multiple distributions as part of the DSL. So e.g. you can easily define product distributions, and ILDJ compat transforms on samples from those products.

Can you do this in the `@dist` DSL? One thing you could do is define a `Product` distribution type -- and then just use the `@dist` DSL with `Product`.

**Comment by femtomc** [comment]
@alex-lew

> For low-dimensional (but still multivariate) invertible transformations, like (x, y) -> (x, x - y), a "triangular" inversion strategy like that used by Stochaskell would be more powerful than the `@dist` DSL is. I think this may be how Jax's program inversion works too? But I am somewhat skeptical of this because it is a bit brittle -- every version I've seen of the idea breaks for (x, y) -> (x + x, y + y), and I believe some break for (x, y) -> (x + y, x - y). So small syntactic variations can break code for somewhat opaque reasons.

The examples provided by your skepticism also break under JAX. 

Here's how the rules work -- `register_binary` adds the rule for the `lax.add_p` primitive to the propagation rules system:

```python
def add_left(left_val, out_val, ildj_):
    return out_val - left_val, ildj_


def add_right(right_val, out_val, ildj_):
    return out_val - right_val, ildj_


register_binary(lax.add_p)(add_left, add_right)
```

Here's `register_binary`:

```python
def register_binary(prim):
    """Registers a binary primitive with ILDJ."""

    def make_rule(f_left, f_right):
        def ildj_rule(incells, outcells, **params):
            (outcell,) = outcells
            left, right = incells
            if outcell.top():
                val, ildj_ = outcell.val, outcell.ildj
                if left.top():
                    right_val, right_ildj = f_left(left.val, val, ildj_)
                    ndslice = NDSlice.new(right_val, right_ildj)
                    incells = [left, InverseAndILDJ(right.aval, [ndslice])]
                elif right.top():
                    left_val, left_ildj = f_right(right.val, val, ildj_)
                    ndslice = NDSlice.new(left_val, left_ildj)
                    incells = [InverseAndILDJ(left.aval, [ndslice]), right]
            elif not outcell.top() and left.top() and right.top():
                out_val = prim.bind(left.val, right.val, **params)
                outcells = [InverseAndILDJ.new(out_val)]
            return incells, outcells, None

        ildj_registry_rules[prim] = ildj_rule

    return make_rule
```

Here `something.top()` is a check for "do I know the value of this node yet?" -- so say we have:

```python
@genjax.dist
def edge_0(x):
    v = genjax.rv(genjax.Normal)(x, 1.0)
    return v + v
```

When I break this program down into `m` and `f` (from above) and try and invert `f` - there's no case to handle when I know the value of the outcell (e.g. that's the return value coming in) and I do not know the value of both the incells (that's `v` and `v`).

@alex-lew Programs like this:

```python
@genjax.dist
def edge_0(x):
    v = genjax.rv(genjax.Normal)(x, 1.0)
    q = genjax.rv(genjax.Normal)(x, 1.0)
    return (v, v - q)
```

do work -- because the propagation runs both forwards and backwards, so it's capable of determining `v`, and then dealing with `v - q`.

But programs like `(x, y) -> (x - y, x + y)` or `(x, y) -> (x + x, y + y)` won't work. For the latter case, you could imagine running a pattern matching interpreter first e.g. see [this pattern matching system in Oryx's experimental area][experimental_matching] (as far as I can tell, this is unused in Oryx - my read on the motivation for this system was to support something similar to [autoconj][autoconj] - that's also "in Oryx, but experimental").

One thing which is really nice about the Oryx code is that even the experimental interpreters / transforms have a understandable comment blurb associated with them.

[experimental_matching]: https://github.com/jax-ml/oryx/blob/main/oryx/experimental/matching/rules.py
[autoconj]: https://arxiv.org/abs/1811.11926

**Comment by femtomc** [comment]
> Nice! But are you sure that you're getting correct densities for this example? jnp.cos is not invertible: the same output can arise from infinitely many different inputs. As such, the density of the transformed distribution is actually an infinite sum of ILDJ-corrected normal densities, evaluated at all the different values where cos(x) = v. This should be intractable, so I'm skeptical that the inversion is being handled properly. Note that with the Prox interfaces, it is valid to just importance sample a single value from the preimage, and compute an importance weight.

As you might have guessed - no I'm not sure :) The rule for `jnp.cos` is `jnp.arccos`. I haven't carefully checked the implementation, I'd assume it is wrong for now.
